### ISSUE TITLE
Bcrypt large speed regression

### ISSUE THREAD

**Issue by Daimanta** [ziglang/zig#13510]
### Zig Version

0.10.0

### Steps to Reproduce and Observed Behavior

The speed of Bcrypt with equal/similar parameters as Zig 0.9.x has regressed by about a factor 10 in 0.10.0. This severely reduces the usability of the implementation as we have this now.

Code as tested:
```
const std2 = @import("std");
const bcrypt = std2.crypto.pwhash.bcrypt;
const time = std2.time;
const begin = time.milliTimestamp();
var params: bcrypt.Params = .{ .rounds_log = 10 };
var hash_options: bcrypt.HashOptions = .{ .allocator = std2.heap.page_allocator, .params = params, .encoding = std2.crypto.pwhash.Encoding.crypt };
var buffer: [bcrypt.hash_length]u8 = undefined;
_ = try bcrypt.strHash("password", hash_options, buffer[0..]);
const end = time.milliTimestamp();
const diff = end - begin;
std2.debug.print("{d}\n", .{diff});
```

This code on my machine has around the following results:
Zig 0.9: 392ms
Zig 0.10: 5128ms

### Expected Behavior

The speed of the algorithm should be similar the speed in Zig 0.9.

**Comment by jedisct1** [comment]
Yes, this is a regression in `stage2` that probably affects more than `bcrypt`.

Compiling with `-fstage1` gets the previous speed back, but this is obviously just a bad workaround.

I'll see if I can pinpoint where the regression is.

**Comment by jedisct1** [comment]
On Apple M1:

```
Stage2: bcrypt:      0.017 s/ops
Stage1: bcrypt:      1.375 s/ops
```

That's a 80x factor.

**Comment by nektro** [comment]
`state: State` vs `state: *const State` and https://github.com/ziglang/zig/issues/12215 could be culprits

**Comment by jedisct1** [comment]
Besides the slowdown:

```
❯ du -sh *.s
280K	main-stage1.s
4.1M	main-stage2.s
```

**Comment by jedisct1** [comment]
> state: State vs state: *const State

That indeed seems to make a massive difference!

**Comment by jedisct1** [comment]
Well done, @nektro !

`state: *const State` indeed brings us back to the stage1 speed.

**Comment by IntegratedQuantum** [comment]
It might be worth noting that as of yesterday related performance problems like #12638 were fixed in #13074
Have you checked with the latest master version if the performance problems still happen?

**Comment by jedisct1** [comment]
Master branch helps, but #13518 still makes a difference:

Apple M1:

```
master: 0.362 s/ops
master+13074: 0.017 s/ops
stage1: 0.017 s/ops
```

**Comment by andrewrk** [comment]
I'm re-opening this issue; it can be closed when 7eed028f9a538624b90279aa430f6e136a77f6a2 is reverted due to fixing the compiler codegen.

**Comment by Daimanta** [comment]
I'm still seeing the same speed issues, although the performance has indeed improved. Running the example code I get around 1100ms for Zig 0.10.1. This is better than 0.10.0 but still about 3 times slower than 0.9.1. I know there is another issue with data passing, but I'd like to keep this issue open as this issue is still heavily impacting the usability of the algorithm.

**Comment by Daimanta** [comment]
Running zig-0.11.0-dev.4003, I have reran the experiment. The speed I'm getting for the above code snippet is 320ms. I would consider this an acceptable speed as it is faster than the 0.9 implementation. I would consider closing this ticket as fixed.

**Comment by jedisct1** [comment]
@Daimanta The core issue doesn't seem to have been resolved.

Reverting https://github.com/ziglang/zig/pull/13518 still causes a significant speed regression.

On a M1 Macbook:

```
Before: 0.014 s/ops
After : 0.364 s/ops
```

**Comment by jedisct1** [comment]
See https://github.com/ziglang/zig/commit/90a877f462fce8bee69ad366aac66805a7c00571 for related issues.
